### PR TITLE
Update "Take Element Screenshot" and "Take Screenshot" user prompt tests to respect unhandled prompt behavior

### PR DESCRIPTION
**Pull request by stanleyhon**
Previously take screenshot and take element screenshot tests were testing that "unhandledPromptBehavior" capability was ignored.

To support this change to the W3C webdriver spec: https://github.com/w3c/webdriver/pull/1405 this change updates the tests to match the spec update.

This change adds:

- check_user_prompt_closed_with_exception, intending to validation prompts are closed but an exception was raised, in line up "dismiss and notify" and "accept and notify" values of unhandledPromptBehavior
- check_user_prompt_automatically_closed, which validates "accept" and "dismiss", where dialogs are closed quietly and without exception
- check_user_prompt_not_modified, which validates dialogs are not interacted with at all.

Testing performed:

- ChromiumEdge and Chromium pass this test.
- Firefox fails this test due to no exception being thrown in cases where we expect exceptions (i.e. dismiss and notify, accept and notify)

### CODE REVIEW

**Comment by andreastt** [comment]
> Firefox fails this test due to no exception being thrown in cases where we expect exceptions (i.e. dismiss and notify, accept and notify)

@whimboo, do you know if we are tracking this?

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (andreastt) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: s/exception/error/g since we probably want to use the WebDriver terminology here.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (andreastt) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: Is there a less expensive command we can run to test that it worked successfully?

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (andreastt) [inline webdriver/tests/take_screenshot/user_prompts.py]: You would probably be able to share this between tests if you put it in https://github.com/web-platform-tests/wpt/blob/master/webdriver/tests/support/fixtures.py.

**Review by andreastt — APPROVED** [review]
I only had some minor comments, but since @whimboo did the original review of https://github.com/w3c/webdriver/pull/1405 it would perhaps be prudent to have him take a look too.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Fixed

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_screenshot/user_prompts.py]: I made some attempts at doing this and ran into some complications:

- take_element_screenshot and take_screenshot differ in that the element variant requires an element ID to take a screenshot of, this complicates the new proposed derived replacement function.
- fixtures.py does not currently have assert functions and inline functions needed by check_user_prompt_closed_with_error, putting this there feels kind of out of place as the other functions seem to be things that help out with tests rather than a generic version of a test.

If you can think of a better way to do this let me know, otherwise I would propose leaving this as is - I think they are different enough to justify a similar copy in two places.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Looking into assert_png, it already does a bare minimum amount of work:

- Decode Base64
- Get and check mime_type.

I don't think there is a direct way to reduce the work here.

I thought about possibly just omitting the assert_png entirely from user_prompts.py files. After all, we are validating the user_prompt behavior around screenshots, not the screenshot itself.

However, I think there is value in keeping them there in case of a browser implementation doing something such as returning success but not successfully taking a screenshot.

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (andreastt) [reply webdriver/tests/take_screenshot/user_prompts.py]: Oh right, thanks for pointing that out. Having one separate for each file makes sense to avoid that complexity.

**Review by andreastt — APPROVED** [review]
This is fine from my point of view, but you probably want a secondary review from someone who knows prompts better.

**Comment by stanleyhon** [comment]
@whimboo looks like the person who authored the file initially, can you provide another review or a merge?

**Comment by whimboo** [comment]
We currently do NOT check for alerts in both of those commands. See the relevant part of the specification: https://w3c.github.io/webdriver/#take-screenshot

@andreastt, as such we would need a decision for https://github.com/w3c/webdriver/issues/1359 first, right? Until then I don't think that we can land this PR.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: Please use the same wording of the fixtures as implemented for any other command. As such we also won't need the comment.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: not needed. this is part of `assert_dialog_handled` which you didn't use here.

**Review by whimboo — CHANGES_REQUESTED** [review]

**Comment by whimboo** [comment]
Just noticed the referenced link to https://github.com/w3c/webdriver/pull/1405 in the initial comment. So this PR is fine and can be landed once reviewed.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Removed these checks.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Reverted to exception terminology and renamed these to match other user_prompt.py files

**Comment by stanleyhon** [comment]
@whimboo I've fixed your feedback up and retested. ready for another look - thanks.

**Comment by stanleyhon** [comment]
@whimboo just a reminder this is ready for another look.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: I would suggest to add a `# META: timeout=long` in the first line, as what we have for all the other user prompt tests. Reason is that those tests can take a bit longer, and would cause timeouts.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: nit: mind listing the imported helper functions in alphabetical order?

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: I miss a call to `assert_dialog_handled()` here. Is there a reason why you aren't checking that?

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: No user prompt tests for all the other commands is doing that. To stay in sync please remove this extra check. Also what you basically want is the call to `assert_dialog_handled()` here too, which plainly is doing such a check that the dialog is gone.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: Please also take the code path of other user prompt tests here, which is checking for the alert text first, and then dismissing the alert. There is no need to run a second command too.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_element_screenshot/user_prompts.py]: nit: you can omit this comment

**Review by whimboo — CHANGES_REQUESTED** [review]
Hey @stanleyhon. Sorry that the review has been taken a bit longer. First I was on PTO and then I wasn't able to catch up with everything else. 

So now I had another look at your patch, and it looks better. But it would still be good if the code paths for each of the tests would be identical to what we have for all the other commands. I gave a couple of replies inline, which would also apply to the tests for `Take Screenshot`.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Done.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Fixed.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Fixed.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Fixed.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: I've reworked most of this change to follow closely what other user_prompts.py files are testing for. Please let me know if you see any other things that may need changing.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Added back assert_dialog_handled everywhere.

**Comment by stanleyhon** [comment]
@whimboo this is ready for another look. I've reworked it to match what other user_prompts.py files do more closely.

CC: @andreastt just FYI.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: I cannot see this entry. Do I miss something?

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_screenshot/user_prompts.py]: Same here for the missing `META` tag for a longer timeout.

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_screenshot/user_prompts.py]: Did you remove this line by accident?

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (whimboo) [inline webdriver/tests/take_screenshot/user_prompts.py]: nit: missing empty line at the end of the file.

**Review by whimboo — CHANGES_REQUESTED** [review]

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: image

 I definitely see it there, but I had to click an icon to expand a collapsed section.

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_screenshot/user_prompts.py]: Same with the first comment you posted, I already see the `META` tag.

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_screenshot/user_prompts.py]: Put it back.

File: webdriver/tests/take_screenshot/user_prompts.py
Comment (stanleyhon) [reply webdriver/tests/take_screenshot/user_prompts.py]: Put it back.

**Comment by stanleyhon** [comment]
@whimboo I've fixed up the new comments. I'm almost certain the `META` tags are in place, I think it's just github UI obscuring it.

File: webdriver/tests/take_element_screenshot/user_prompts.py
Comment (whimboo) [reply webdriver/tests/take_element_screenshot/user_prompts.py]: Not that easy to see. Thanks for highlighting that. Expanding this comment section brings-up those meta tags.

**Review by whimboo — APPROVED** [review]
Thanks! It looks good now, so lets get it merged. \o/

**Comment by andreastt** [comment]
@whimboo I didn’t look too carefully at this, but did you see https://bugzilla.mozilla.org/show_bug.cgi?id=1583401#c5?

Are these expected changes?

**Comment by whimboo** [comment]
Sure. That's why I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1584927 because Marionette doesn't have it implemented yet.

**Comment by andreastt** [comment]
Perfect, I didn’t see that yet! 👍🏻

**Pull request merged**
